File: V2/quest-client/src/rendering/LatencyEstimator.ts
/**
 * Per-frame latency estimation from the server's `pts` media timestamps.
 *
 * Absolute capture→arrival latency cannot be derived from pts alone (the RTP
 * clock has an arbitrary offset from the client clock), but its *variation*
 * can: `offset_i = arrival_i − pts_i(ms)` is constant when frames flow
 * smoothly and grows when a frame queued behind slow inference or network
 * jitter. The estimate is therefore
 *
 *     latency_i = baseLatencyMs + (offset_i − min(offset over recent window))
 *
 * — the configured baseline plus the measured queuing delay. The windowed
 * minimum re-anchors the baseline as conditions drift.
 */

const RTP_CLOCK_HZ = 90_000;

export class LatencyEstimator {
  /** `[arrivalMs, offsetMs]` pairs within the window, arrival-ordered. */
  private readonly offsets: Array<[number, number]> = [];
  private lastOffsetMs: number | undefined;

  constructor(
    private readonly baseLatencyMs = 200,
    private readonly windowMs = 5000,
  ) {}

  observe(arrivalMs: number, pts: number | null | undefined): void {
    if (typeof pts !== 'number' || !Number.isFinite(pts)) return;
    const offsetMs = arrivalMs - (pts / RTP_CLOCK_HZ) * 1000;
    this.lastOffsetMs = offsetMs;
    this.offsets.push([arrivalMs, offsetMs]);
    const cutoff = arrivalMs - this.windowMs;
    while (this.offsets.length > 1 && this.offsets[0]![0] < cutoff) {
      this.offsets.shift();
    }
  }

  latencyMs(): number {
    if (this.lastOffsetMs === undefined || this.offsets.length === 0) {
      return this.baseLatencyMs;
    }
    let minOffset = Number.POSITIVE_INFINITY;
    for (const [, offset] of this.offsets) {
      if (offset < minOffset) minOffset = offset;
    }
    return this.baseLatencyMs + Math.max(0, this.lastOffsetMs - minOffset);
  }
}

File: V2/com.questvisionstream.unity/Runtime/Core/LatencyEstimator.cs
// Copyright (c) Simon Jackson (SimonDarksideJ). All rights reserved.
// Licensed under the MIT License. See LICENSE in the repository root for license information.

using System.Collections.Generic;

namespace QuestVisionStream.Core
{
    /// <summary>
    /// Per-session latency estimation from the server's <c>pts</c> media timestamps.
    ///
    /// Absolute capture-to-arrival latency cannot be derived from pts alone (the RTP
    /// clock has an arbitrary offset from the client clock), but its variation can:
    /// <c>offset_i = arrival_i − pts_i(ms)</c> is constant when frames flow smoothly
    /// and grows when a frame queued behind slow inference or network jitter. So:
    /// <code>latency_i = baseLatencyMs + (offset_i − min(offset over recent window))</code>
    /// — the configured baseline plus the measured queuing delay. The windowed
    /// minimum re-anchors the baseline as conditions drift. Frames without pts fall
    /// back to the base. Only a clock-synchronized protocol could remove the base.
    /// </summary>
    public sealed class LatencyEstimator
    {
        private const double RtpClockHz = 90_000;

        /// <summary>(arrivalMs, offsetMs) pairs within the window, arrival-ordered.</summary>
        private readonly List<(double arrivalMs, double offsetMs)> offsets = new List<(double, double)>();
        private readonly double baseLatencyMs;
        private readonly double windowMs;
        private double? lastOffsetMs;

        public LatencyEstimator(double baseLatencyMs = 200, double windowMs = 5000)
        {
            this.baseLatencyMs = baseLatencyMs;
            this.windowMs = windowMs;
        }

        public void Observe(double arrivalMs, long? pts)
        {
            if (!pts.HasValue)
            {
                return;
            }

            var offsetMs = arrivalMs - (pts.Value / RtpClockHz) * 1000.0;
            lastOffsetMs = offsetMs;
            offsets.Add((arrivalMs, offsetMs));

            var cutoff = arrivalMs - windowMs;
            while (offsets.Count > 1 && offsets[0].arrivalMs < cutoff)
            {
                offsets.RemoveAt(0);
            }
        }

        public double LatencyMs()
        {
            if (!lastOffsetMs.HasValue || offsets.Count == 0)
            {
                return baseLatencyMs;
            }

            var minOffset = double.PositiveInfinity;
            foreach (var (_, offset) in offsets)
            {
                if (offset < minOffset)
                {
                    minOffset = offset;
                }
            }

            return baseLatencyMs + System.Math.Max(0, lastOffsetMs.Value - minOffset);
        }

        public void Reset()
        {
            offsets.Clear();
            lastOffsetMs = null;
        }
    }
}
